### PR TITLE
[Serving] Implement reject sampling algorithm for top-p sampling

### PR DESCRIPTION
**Pull request by yzh119**
Our current implementation of top-p sampling relies on sorting, which takes O(n log n) time, and not efficient enough.
This PR implements a reject sampling algorithm for top-p sampling:
1. draw sample i from the given probability distribution
2. compute sum_{j, p_j < p_i} (p_j), if the sum >= 1 - top_p, it indicates p_i is within top_p, accept the draw and return i.
3. else, reject i and mask all probabilities p_j with 0 if p_j < p_i, and re-normalize the probability distributions and top_p, go to step 1

The expected number of rounds is around O(1/p), we set a maximum number of rejects N, if the algorithm cannot find eligible samples within N, use sorting-based sampling instead.

Unittests are not implemented yet, do not merge this PR before that.

### CODE REVIEW

File: cpp/serve/sampler.cc [hunk 1]
```diff
@@ -118,6 +118,95 @@ void SetProbWithArgmaxOnCPU(NDArray logits, int token_offset) {
...
+        if (normalized_prob[j] < pi) {
+          normalized_prob[j] = 0.f;
+        } else {
+          normalized_prob[j] *= normalize_scale;
+        }
+      }
+    }
+  }
+
+  // cannot find eligible choice in max_num_rejects rounds, return -1
```
Comment (tqchen) [inline cpp/serve/sampler.cc:206]: Let us simply fold the normal sample and reject sample part as a single function and keep those two as internal sub routines

File: cpp/serve/sampler.cc [hunk 2]
```diff
@@ -118,6 +118,95 @@ void SetProbWithArgmaxOnCPU(NDArray logits, int token_offset) {
...
+  ICHECK(prob->device.device_type == kDLCPU);
+
+  int64_t ndata = prob->shape[prob->ndim - 1];
+  const float* __restrict p_prob =
+      static_cast<float*>(__builtin_assume_aligned(prob->data, 4)) + (unit_offset * ndata);
+  constexpr double one = 1.0f - 1e-5f;
+
+  auto draw_sample_from_prob_distribution = [&](const float* p, size_t ndata,
+                                                double uniform_sample) {
+    double prob_sum = 0.f;
```
Comment (tqchen) [inline cpp/serve/sampler.cc:159]: need to confirm speed of double sum, likely float32 is sufficient

File: cpp/serve/sampler.cc [hunk 3]
```diff
@@ -118,6 +118,95 @@ void SetProbWithArgmaxOnCPU(NDArray logits, int token_offset) {
...
+  };
+
+  if (top_p >= one) {
+    // Specially handle case where top_p == 1.
+    return draw_sample_from_prob_distribution(p_prob, ndata, uniform_samples[0]);
+  }
+  double q = 1 - top_p;
+
+  // rejection sampling
+  std::vector<float> normalized_prob(p_prob, p_prob + ndata);
```
Comment (tqchen) [inline cpp/serve/sampler.cc:177]: no need to renormalize probablities, we can simply redo it

File: cpp/serve/sampler.cc [hunk 4]
```diff
@@ -118,6 +118,95 @@ void SetProbWithArgmaxOnCPU(NDArray logits, int token_offset) {
...
+
+  const int max_num_rejects = uniform_samples.size();
+  for (int round = 0; round < max_num_rejects; ++round) {
+    int i =
+        draw_sample_from_prob_distribution(normalized_prob.data(), ndata, uniform_samples[round]);
+    float pi = normalized_prob[i];
+    double prob_sum_lt_pi = 0.f;
+    for (int j = 0; j < ndata; ++j) {
+      if (normalized_prob[j] < pi) {
```
Comment (tqchen) [inline cpp/serve/sampler.cc:186]: we need to do early exit.

keep two running sums:
- prob_sum_lt_pi
- 
```
double prob_sum_gt_pi = 0;
double prob_sum_le_pi = pi;
double running_sum = pi;

for (j) {
   if (j != selected_index) {
     running_sum += p_prob[j];
     if (p_prob[j] > pi) {
        prob_sum_gt_pi += p_prob[j];
     } else {
        prob_sum_le_pi += p_prob[j];
     }
   }
   // condition checking 
   if (1 - running_sum <= pi) {
      // early exit, run check condition below
      sum_le_pi += 1 - running_sum;
   } 
   // check condition
   if (sum_le_pi >= 1 - top_p) {
      // early accept
   } else if (sum_gt_pi >= top_p) {
      // early reject 
   }
}
```

File: cpp/serve/sampler.cc [hunk 5]
```diff
@@ -118,6 +118,95 @@ void SetProbWithArgmaxOnCPU(NDArray logits, int token_offset) {
...
+ * \param top_p The top-p value of sampling.
+ * \param uniform_sample A list of random numbers in [0, 1] for sampling, the
+ * size of the list is the maximum number of rejects we can tolerate for reject
+ * sampling.
+ * \return The sampled value.
+ * \note If we cannot find eligible choice within max_num_rejects rounds, this
+ * function will return -1.
+ */
+int RejectSampleTopPFromProb(NDArray prob, int unit_offset, double top_p,
+                             const std::vector<double>& uniform_samples) {
```
Comment (tqchen) [inline cpp/serve/sampler.cc:136]: consider take in uniform samples as NDArray same as prob now that we take in the data.

File: cpp/serve/sampler.cc [hunk 6]
```diff
@@ -391,20 +480,33 @@ class CPUSampler : public SamplerObj {
     CHECK_EQ(probs->ndim, 2);
     CHECK_EQ(probs->device.device_type, kDLCPU);
 
+    const int max_num_rejects = 6;
     int n = probs->shape[0];
-    std::vector<double> random_numbers;
+    std::vector<std::vector<double>> random_numbers;
```
Comment (tqchen) [inline cpp/serve/sampler.cc:485]: allocate a NDArray that can be reused across runs, and populate data

File: cpp/serve/sampler.cc [hunk 7]
```diff
@@ -391,20 +480,33 @@ class CPUSampler : public SamplerObj {
...
+      random_numbers.push_back(random_numbers_i);
     }
 
     tvm::runtime::parallel_for_with_threading_backend(
         [&sampled_tokens, &probs, &generation_cfg, &random_numbers](int i) {
           // Sample top p from probability.
-          sampled_tokens[i] =
-              SampleTopPFromProb(probs, i, generation_cfg[i]->top_p, random_numbers[i]);
+          int reject_sampled_token =
+              RejectSampleTopPFromProb(probs, i, generation_cfg[i]->top_p, random_numbers[i]);
```
Comment (tqchen) [inline cpp/serve/sampler.cc:501]: combine the two sampling function together as sub functions functions of a global sample function that takes in prob and NDArray of uniform

**Review by tqchen — CHANGES_REQUESTED** [review]

File: cpp/serve/sampler.cc [hunk 8]
```diff
@@ -118,6 +118,95 @@ void SetProbWithArgmaxOnCPU(NDArray logits, int token_offset) {
...
+ * \param unit_offset The offset specifying which distribution to sample from.
+ * \param top_p The top-p value of sampling.
+ * \param uniform_sample A list of random numbers in [0, 1] for sampling, the
+ * size of the list is the maximum number of rejects we can tolerate for reject
+ * sampling.
+ * \return The sampled value.
+ * \note If we cannot find eligible choice within max_num_rejects rounds, this
+ * function will return -1.
+ */
+int RejectSampleTopPFromProb(NDArray prob, int unit_offset, double top_p,
```
Comment (tqchen) [inline cpp/serve/sampler.cc:135]: consider the following structure

```c++
// one round of reject sampling, no renormalization to avoid write
int TryRejectSampleTopPFromProb(NDArray prob, int unit_offset, float top_p, float uniform_sample) {
}
// The uniform sample NDArray is n * num_max_sample, we do this so that we can do deterministic
// parallelization and pre-populate the uniform_samples before parallel sampling
int SampleTopPFromProb(NDArray prob, int unit_offset, float top_p, NDArray uniform_samples) {
}
```

**Review by tqchen — CHANGES_REQUESTED** [review]
